### PR TITLE
[RFC] Refactor container flushing logic

### PR DESCRIPTION
**Pull request by zamotany**
* Better perf
* Base ground for further optimizations

TODO:
* Profile and determinate if it's really worth merging

cc: @satya164

### CODE REVIEW

File: src/nodes/ContainerNode.js [hunk 1]
```diff
@@ -37,79 +39,104 @@ export default class ContainerNode {
...
+      const lines = chunk.data.split('\n');
+      const hasNewLines = chunk.data.includes('\n');
+
+      if (!hasNewLines) {
+        normalizedBuffer.push(chunk);
+      } else {
+        lines
+          .slice(0, lines.length - 1)
+          .map(line => (line === '' ? '\n' : line))
+          .forEach(line => {
```
Comment (thymikee) [inline src/nodes/ContainerNode.js]: `lines` are not used anywhere else, so you can use just one `forEach`:
```js
lines
  .slice(0, lines.length - 1)
  .forEach(line => {
    normalizedBuffer.push({ data: line === '' ? '\n' : line });
  });
```

File: example/src/listViewer/index.js [hunk 2]
```diff
@@ -96,10 +96,26 @@ class Viewer extends React.Component {
   }
 }
 
+class Throttle extends React.Component {
```
Comment (thymikee) [inline example/src/listViewer/index.js]: A leftover or is it intentionally here?

Comment (zamotany) [reply example/src/listViewer/index.js]: leftover

File: src/nodes/ContainerNode.js (hunk 1 above)
Comment (satya164) [reply src/nodes/ContainerNode.js]: or

```js
normalizedBuffer.push(
  ...(
    lines.slice(0, lines.length - 1)
      .map(line => ({ data: line === '' ? '\n' : line }))
  )
)
```

¯\_(ツ)_/¯

**Comment by satya164** [comment]
I'll take a good look soon after I take some rest :D

Thank you!

File: src/nodes/ContainerNode.js (hunk 1 above)
Comment (thymikee) [reply src/nodes/ContainerNode.js]: oh, and you can just use `.slice(0, -1)`

File: src/nodes/ContainerNode.js [hunk 3]
```diff
@@ -2,27 +2,29 @@
...
 
+type Chunk = {
+  data: string,
+};
+
 export default class ContainerNode {
   children = [];
-  frontBuffer: string = '';
-  backBuffer: string = '';
+  frontBuffer: Chunk[] = [];
```
Comment (satya164) [inline src/nodes/ContainerNode.js]: maybe nicer to name them as `previousBuffer` and `nextBuffer`?

**Comment by satya164** [comment]
So basically,

1. If the number of lines hasn't changed, we clear and write the changed ones
2. If the number of lines has changed, we clear everything below the point of change and write again

Looks good.

**Comment by zamotany** [comment]
Yes, thats the idea but I'm not entirely sure if it worth complicating the logic, performance gain is unnoticeable.

**Comment by satya164** [comment]
Makes sense. What about the CPU usage? There's no decrease?

**Comment by zamotany** [comment]
Depends on what is happening. If only one spinner is running with 200ms interval the CPU is running on 5%, but when I start to interact, pressing W/S keys so the blue highlighting starts to move I got it to max 20%, but only for few seconds, then it drops. This is connected with rendering the whole component tree.

**Comment by satya164** [comment]
Earlier we had high CPU usage with only spinner, right?

**Comment by zamotany** [comment]
I don't think so, but I don't remember actually.

**Comment by zamotany** [comment]
And this approach might perform even worse in bigger apps. Cause for most of the time we need to iterate every line and with previous approach we could bail out on first mismatch.

**Comment by satya164** [comment]
Makes sense :)

**Comment by zamotany** [comment]
Closing this as it's no longer relevant.

**Pull request closed without merging**